src/EventHandler.ts

```diff
@@ -0,0 +1,248 @@
+export default class EventHandler {
+    private pendingMouseEventStart: MouseEvent | null;
+    private pendingMouseEventEnd: MouseEvent | null;
+    private pendingTouchEventStart: TouchEvent | null;
+    private pendingTouchEventEnd: TouchEvent | null;
+
+    public onLeftTap: (event: UIEvent) => void = () => {};
+    public onMiddleTap: (event: UIEvent) => void = () => {};
+    public onRightTap: (event: UIEvent) => void = () => {};
+    public onBackwardSwipe: (event: UIEvent) => void = () => {};
+    public onForwardSwipe: (event: UIEvent) => void = () => {};
+
+    public onLeftHover: () => void = () => {};
+    public onRightHover: () => void = () => {};
+    public onRemoveHover: () => void = () => {};
+
+    public constructor() {
+        this.pendingMouseEventStart = null;
+        this.pendingMouseEventEnd = null;
+        this.pendingTouchEventStart = null;
+        this.pendingTouchEventEnd = null;
+    }
+
+    public setupEvents(element: HTMLElement | Document) {
+        if (this.isTouchDevice()) {
+            element.addEventListener("touchstart", this.handleTouchEventStart.bind(this));
+            element.addEventListener("touchend", this.handleTouchEventEnd.bind(this));
+        } else {
+            element.addEventListener("mousedown", this.handleMouseEventStart.bind(this));
+            element.addEventListener("mouseup", this.handleMouseEventEnd.bind(this));
+            element.addEventListener("mouseenter", this.handleMouseMove.bind(this));
+            element.addEventListener("mousemove", this.handleMouseMove.bind(this));
+            element.addEventListener("mouseleave", this.handleMouseLeave.bind(this));
+        }
+    }
+
+    private isTouchDevice() {
+        return !!('ontouchstart' in window || navigator.maxTouchPoints);
+    }
+
+    private handleMouseEventStart = (event: MouseEvent): void => {
+        this.pendingMouseEventStart = event;
+    }
+
+    private handleTouchEventStart = (event: TouchEvent): void => {
+        if (event.changedTouches.length !== 1) {
+            // This is a multi-touch event. Ignore.
+            return;
+        }
+
+        this.pendingTouchEventStart = event;
+    }
+
+    private handleMouseEventEnd = (event: MouseEvent): void => {
+        if (!this.pendingMouseEventStart) {
+            // Somehow we got an end event without a start event. Ignore it.
```

Comment:
I must've had something scrambled in my brain when I wrote that comment. Please ignore. ^_^